sky/serve/replica_managers.py

```diff
@@ -645,25 +725,64 @@ def _download_and_stream_logs(info: ReplicaInfo):
                                                      log_file_name, 'a').run,
-            args=(info.cluster_name,),
+            args=(info.cluster_name, delay_in_seconds),
         )
         info.status_property.sky_down_status = ProcessStatus.RUNNING
         serve_state.add_or_update_replica(self._service_name, replica_id, info)
         p.start()
         self._down_process_pool[replica_id] = p
 
     def scale_down(self, replica_id: int) -> None:
-        self._terminate_replica(replica_id, sync_down_logs=False)
+        self._terminate_replica(replica_id,
+                                sync_down_logs=False,
+                                delay_in_seconds=_DEFAULT_DRAIN_SECONDS)
 
-    def _handle_preemption(self, replica_id: int) -> None:
-        logger.info(f'Beginning handle for preempted replica {replica_id}.')
-        # TODO(MaoZiming): Support spot recovery policies
-        info = serve_state.get_replica_info_from_id(self._service_name,
-                                                    replica_id)
-        assert info is not None
+    def _handle_preemption(self, info: ReplicaInfo) -> bool:
+        """Handle preemption of the replica if any error happened.
+
+        Returns:
+            bool: Whether the replica is preempted.
+        """
+        if not info.is_spot:
+            return False
+
+        # Get cluster handle first for zone information. The following
+        # backend_utils.refresh_cluster_status_handle might delete the
+        # cluster record from the cluster table.
+        handle = global_user_state.get_handle_from_cluster_name(
+            info.cluster_name)
+        if handle is None:
+            logger.error(f'Cannot find cluster {info.cluster_name} for '
+                         f'replica {info.replica_id} in the cluster table. '
+                         'Skipping preemption handling.')
+            return False
+        assert isinstance(handle, backends.CloudVmRayResourceHandle)
+        # Pull the actual cluster status from the cloud provider to
+        # determine whether the cluster is preempted.
+        cluster_status, _ = backend_utils.refresh_cluster_status_handle(
+            info.cluster_name,
+            force_refresh_statuses=set(status_lib.ClusterStatus))
+
+        if cluster_status == status_lib.ClusterStatus.UP:
+            return False
+        # The cluster is (partially) preempted. It can be down, INIT or STOPPED,
+        # based on the interruption behavior of the cloud.
+        cluster_status_str = ('' if cluster_status is None else
+                              f' (status: {cluster_status.value})')
+        logger.info(
+            f'Replica {info.replica_id} is preempted{cluster_status_str}.')
+        if info.location is None:
+            logger.error(f'Cannot find location for replica {info.replica_id}. '
+                         'Skipping adding to preemption list.')
+        else:
+            self.preemption_history.append(info.location)
```

Comment:
We should only do this for spot policy service. Otherwise it is a list that keeps expanding in memory (since we only cleanup this list if spot policy is used)